lib/Dialect/TT/IR/TTOpsTypes.cpp

```diff
@@ -513,8 +511,79 @@ MemorySpace LayoutAttr::getMemorySpace() const {
       .getValue();
 }
 
-mlir::AffineMap LayoutAttr::projectOnto(ArrayRef<int64_t> logicalTensorShape,
-                                        GridAttr grid) const {
+// Returns shape of the tensor after tilization is applied.
+llvm::SmallVector<int64_t>
+LayoutAttr::getTiledShape(llvm::ArrayRef<int64_t> tensorShape) const {
+  assert(isTiled() && "Expected a tiled layout");
+
+  auto tileLinearMap = getTileLinearMap();
+
+  mlir::SmallVector<mlir::AffineExpr> tensorShapeExprs(
+      llvm::map_range(tensorShape, [context = getContext()](std::int64_t e) {
+        return getAffineConstantExpr(e - 1, context);
+      }));
+
+  mlir::SmallVector<std::int64_t> tiledShape(tileLinearMap.getNumResults());
+  for (unsigned i = 0; i < tileLinearMap.getNumResults(); ++i) {
+    mlir::AffineExpr expr = tileLinearMap.getResult(i);
+    mlir::AffineExpr constantExpr = expr.replaceDims(tensorShapeExprs);
+    tiledShape[i] =
+        llvm::cast<mlir::AffineConstantExpr>(constantExpr).getValue() + 1;
+  }
+
+  return tiledShape;
+}
+
+// Returns an affine map that maps tensor original dimensions to the tile linear
+// space instead of scalar space.
+// Example:
+// scalar linear: (d0, d1, d2, d3) -> (d0 * 192 + d1 * 64 + d2, d3),
+// tile linear: (d0, d1, d2, d3) -> (d0 * 6 + d1 * 2 + d2 floordiv 32, d3
+// floordiv 32).
+mlir::AffineMap LayoutAttr::getTileLinearMap() const {
+  assert(isTiled() && "Expected a tiled layout");
+
+  mlir::AffineMap linear = getLinear();
+  mlir::SmallVector<AffineExpr> tileLinearExprs;
+
+  TileType tileType = mlir::cast<TileType>(getElementType());
+  int64_t tileH = tileType.getHeight();
+  int64_t tileW = tileType.getWidth();
+
+  mlir::AffineExpr d0 = getAffineDimExpr(0, getContext());
+  mlir::AffineExpr d1 = getAffineDimExpr(1, getContext());
+  tileLinearExprs.push_back(d0.floorDiv(tileH));
+  tileLinearExprs.push_back(d1.floorDiv(tileW));
+
+  mlir::AffineMap tilingMap =
+      AffineMap::get(2 /* tile rank */, 0, tileLinearExprs, getContext());
+
+  // If the linear map has more than two results, drop the first results and
+  // leave the last two. This is needed before composition with the tiling map.
+  // The dropped results are added back after composition.
+  mlir::SmallVector<int64_t> dropPositions;
+  mlir::SmallVector<AffineExpr> droppedResults;
+  for (uint i = 0; i < linear.getNumResults() - 2; ++i) {
```

Comment:
We should probably use signed ints here.  If linear.getNumResults() is < 2 we'll underflow.  In this PR I just added support for 1d tensors so it might be possible to get < 2 num results #478.

For now it might be OK to assert that `linear.getNumResults() >= 2` if we don't want to deal with the 1d case yet here.

Reply:
Good catch, compiler was complaining for sign comparison, so I put `uint` not thinking much about it. Here we need at least 2 results in `linear` for compose op to work. I am not sure how we'd translate 1D linear to 2D tiled shape, so maybe assert is the way to go for now.

Reply:
Yeah I think assert makes sense